File: src/main/java/com/hepdd/gtmthings/api/gui/widget/SimpleNumberInputWidget.java
package com.hepdd.gtmthings.api.gui.widget;

import net.minecraft.network.FriendlyByteBuf;
import net.minecraft.util.Mth;

import com.lowdragmc.lowdraglib.gui.widget.TextFieldWidget;
import com.lowdragmc.lowdraglib.gui.widget.WidgetGroup;
import lombok.Getter;

import java.util.function.IntConsumer;
import java.util.function.IntSupplier;

public class SimpleNumberInputWidget extends WidgetGroup {

    @Getter
    private final IntSupplier valueSupplier;
    @Getter
    private int min = defaultMin();
    @Getter
    private int max = defaultMax();

    private final IntConsumer onChanged;

    private TextFieldWidget textField;

    protected String toText(int value) {
        return String.valueOf(value);
    }

    protected int fromText(String value) {
        return Integer.parseInt(value);
    }

    protected int clamp(int value, int min, int max) {
        return Mth.clamp(value, min, max);
    }

    protected int defaultMin() {
        return 0;
    }

    protected int defaultMax() {
        return Integer.MAX_VALUE;
    }

    protected void setTextFieldRange(TextFieldWidget textField, int min, int max) {
        textField.setNumbersOnly(min, max);
    }

    public SimpleNumberInputWidget(int x, int y, int width, int height, IntSupplier valueSupplier,
                                   IntConsumer onChanged) {
        super(x, y, width, height);
        this.valueSupplier = valueSupplier;
        this.onChanged = onChanged;
        buildUI();
    }

    @Override
    public void initWidget() {
        super.initWidget();
        this.textField.setCurrentString(toText(valueSupplier.getAsInt()));
    }

    @Override
    public void writeInitialData(FriendlyByteBuf buffer) {
        super.writeInitialData(buffer);
        buffer.writeUtf(toText(valueSupplier.getAsInt()));
    }

    @Override
    public void readInitialData(FriendlyByteBuf buffer) {
        super.readInitialData(buffer);
        this.textField.setCurrentString(buffer.readUtf());
    }

    private void buildUI() {
        this.textField = new TextFieldWidget(0, 0, getSizeWidth(), 12,
                () -> toText(valueSupplier.getAsInt()),
                stringValue -> this.setValue(clamp(fromText(stringValue), min, max)));

        this.addWidget(this.textField);
    }

    public SimpleNumberInputWidget setValue(int value) {
        onChanged.accept(value);
        return this;
    }

    public SimpleNumberInputWidget setMin(int min) {
        this.min = min;
        updateTextFieldRange();

        return this;
    }

    public SimpleNumberInputWidget setMax(int max) {
        this.max = max;
        updateTextFieldRange();

        return this;
    }

    protected void updateTextFieldRange() {
        setTextFieldRange(textField, min, max);

        this.setValue(clamp(valueSupplier.getAsInt(), min, max));
    }
}

File: src/main/java/com/hepdd/gtmthings/api/gui/widget/SimpleNumberInputWidget.kt
package com.hepdd.gtmthings.api.gui.widget

import net.minecraft.network.FriendlyByteBuf
import net.minecraft.util.Mth

import com.lowdragmc.lowdraglib.gui.widget.TextFieldWidget
import com.lowdragmc.lowdraglib.gui.widget.WidgetGroup

import java.util.function.Consumer
import java.util.function.Supplier

open class SimpleNumberInputWidget(x: Int, y: Int, width: Int, height: Int, valueSupplier: Supplier<Int?>, onChanged: Consumer<Int?>) : WidgetGroup(x, y, width, height) {
    private var valueSupplier: Supplier<Int?>? = null

    private var min: Int = defaultMin()

    private var max: Int = defaultMax()

    private var onChanged: Consumer<Int?>? = null

    private var textField: TextFieldWidget? = null

    init {
        this.valueSupplier = valueSupplier
        this.onChanged = onChanged
        buildUI()
    }

    private fun toText(value: Int?): String? = value.toString()

    private fun fromText(value: String): Int = value.toInt()

    private fun clamp(value: Int, min: Int, max: Int): Int = Mth.clamp(value, min, max)

    private fun defaultMin(): Int = 0

    private fun defaultMax(): Int = Int.Companion.MAX_VALUE

    private fun setTextFieldRange(textField: TextFieldWidget, min: Int, max: Int) {
        textField.setNumbersOnly(min, max)
    }

    override fun initWidget() {
        super.initWidget()
        this.textField!!.setCurrentString(toText(valueSupplier!!.get()))
    }

    override fun writeInitialData(buffer: FriendlyByteBuf) {
        super.writeInitialData(buffer)
        buffer.writeUtf(toText(valueSupplier!!.get())!!)
    }

    override fun readInitialData(buffer: FriendlyByteBuf) {
        super.readInitialData(buffer)
        this.textField!!.setCurrentString(buffer.readUtf())
    }

    private fun buildUI() {
        this.textField = TextFieldWidget(
            0,
            0,
            sizeWidth,
            12,
            { toText(valueSupplier!!.get()) },
            { stringValue: String? -> this.setValue(clamp(fromText(stringValue!!), min, max)) },
        )

        this.addWidget(this.textField)
    }

    fun setValue(value: Int?): SimpleNumberInputWidget {
        onChanged!!.accept(value)
        return this
    }

    fun setMin(min: Int): SimpleNumberInputWidget {
        this.min = min
        updateTextFieldRange()

        return this
    }

    fun setMax(max: Int): SimpleNumberInputWidget {
        this.max = max
        updateTextFieldRange()

        return this
    }

    protected fun updateTextFieldRange() {
        setTextFieldRange(textField!!, min, max)

        this.setValue(clamp(valueSupplier!!.get()!!, min, max))
    }
}
